proposals/p2029.md

```diff
@@ -0,0 +1,103 @@
+# C++ Interop: Reserved identifiers
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+[Pull request](https://github.com/carbon-language/carbon-lang/pull/2029)
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Problem](#problem)
+-   [Background](#background)
+-   [Proposal](#proposal)
+-   [Details](#details)
+-   [Rationale](#rationale)
+-   [Alternatives considered](#alternatives-considered)
+
+<!-- tocstop -->
+
+## Problem
+
+C++ reserves certain sets of identifiers for implementation-specific use. Carbon
+must ensure it respects these reserved identifiers to enable seamless and
+error-free interoperability with C++. Equally, it would be prudent to define a
+set of identifiers which are reserved for internal Carbon implementation usage.
+
+## Background
+
+The full set of C++ reserved identifiers is listed
+[here](https://en.cppreference.com/w/cpp/language/identifiers). This proposal is
+solely considering the set of user-definable identifiers which are considered
+names in C++:
+
+-   Identifiers with a double underscore anywhere are reserved;
+-   Identifiers that begin with an underscore followed by an uppercase letter
+    are reserved;
+-   Identifiers that begin with an underscore are reserved in the global
+    namespace.
+
+Considerations for reserved keywords are discussed in
+[p0093](https://github.com/carbon-language/carbon-lang/pull/93).
+
+## Proposal
+
+Carbon reserves the following identifiers in the specified contexts:
+
+-   Identifiers with a double underscore anywhere are reserved;
+-   Identifiers that begin with an underscore followed by an uppercase letter
+    are reserved.
+
+Carbon toolchains are expected to emit an error when a reserved identifier is
+encountered. This will remove the risk of undefined / unexpected runtime
+behaviour, or name resolution clashes when linking with C++ libraries.
+
+## Details
+
+This proposal makes use of three assumptions:
+
+1. A Carbon package exporting a C++ header will be scoped in a C++ namespace
+   following the package name;
+2. All C++ imports to Carbon will exist within the package `Cpp` which
+   represents the root of the C++ global namespace.
+3. The default Carbon package can not be exported to C++, following the
+   [no other library can import this library even from within the default package](/docs/design/README.md#files-libraries-packages)
+   rule.
```

Comment:
I think this is a misunderstanding. As written, I think it's confusing packages and libraries, and maybe you meant to say:

"_A package's default library_ can not be exported to C++, following the [no other library can import _the default library_ even from within the _same package_](/docs/design/README.md#files-libraries-packages) rule."

As presently discussed, no other library _within the current package_ can import the default library. That is, a library written `package Foo api;` cannot be imported from elsewhere within the package `Foo` because `import Foo;` would be the correct syntax, but `import library <name>` is _required_ when a library imports from itself.

In other words, other packages can import the default library of a package. And similarly, the default library can be exported to C++.

Reply:
I was referring to these two elements of the linked doc, which are explicitly about default package export or import (although the second point does link the two defaults):

- If the package name is omitted, as in `package library "Main" api`;, the file contributes to the default package. No other package may import from the default package.
- If a file has no package declaration at all, it is the `api` file belonging to the default package and default library. This is particularly for tests and smaller examples. No other library can import this library even from within the default package. It can be split across multiple `impl` files using a `package impl;` package declaration.

I was thinking about whether an un-named package could be exported to C++, and what namespace implications that could have (i.e. if it were possible, would it be to the root of the global namespace, and therefore do we need to worry about collissions?) Grateful to know if I have misunderstood something here?

Reply:
Thanks! I'd missed that this was added to the readme. It's provisional and not really part of the design, so I haven't thought about it much -- you're probably right.